### DISCUSSION TITLE
UI framework

### DISCUSSION THREAD

**Discussion by erengy** [erengy/taiga#1050] [Ideas]
Taiga has been a native Windows application since day one (16d280fb90ec061e8d1eb5215116c8c5e828f176), using [Win32 API](https://docs.microsoft.com/en-us/windows/win32/). Eventually most of the relevant code was moved to a standalone [library](https://github.com/erengy/windows) (#316), but the underlying technology has not changed.

Native look and behavior is one of the main characteristics of Taiga. But we're stuck with Windows (#92), and working on UI-related improvements is a pain (e.g. #807, #533, #87). While it's going to take quite an effort, switching to a different UI framework is necessary for going cross-platform and getting any serious UI work done.

Below is a list of alternatives, each with their own pros and cons. First step is to evaluate them thoroughly and decide on which one to use.

## [wxWidgets](https://www.wxwidgets.org/)

wxWidgets does what Taiga does, only better: Cross-platform, truly native controls, lightweight, allows using just the event system and GUI classes without becoming a huge dependency. Feels like it would play well with other parts of the application and overall require the least amount of work. However, some of the limitations Taiga currently has would carry over. For instance, we cannot have dark mode on Windows until the OS itself officially supports it for Win32 apps.

An application similar to Taiga is [PicoTorrent](https://picotorrent.org/). Started with Win32 API, [switched](https://picotorrent.blog/behind-the-release-v019/) to WTL to wxWidgets to Qt, then back to wxWidgets again. Maybe @vktr can chime in to tell if they're still content with it. A counter example would be [Dolphin Emulator](https://dolphin-emu.org/), which [moved](https://dolphin-emu.org/blog/2017/06/03/dolphin-progress-report-may-2017/#qt-ie-patootie) from wxWidgets to Qt. Looks like they *really* wanted to [leave](https://dolphin-emu.org/blog/2018/05/02/legend-dolphin-lens-between-worlds/) wxWidgets behind.

## [Qt](https://www.qt.io/)

Qt is arguably the highest quality and most powerful framework out there, with great documentation. While it offers native look and feel, it doesn't create actual native controls, which allows for easier customization (e.g. dark theme). Qt is much more than a UI framework though, and it [comes](https://www.reddit.com/r/cpp/comments/5t7d6u/trying_qt_seems_great_what_downsides_am_i_missing/ddkswoz/?context=3) with non-standard features and conventions. If we decide on using Qt, it's going to be more intrusive than any other option.

Qt 6 [requires](https://doc.qt.io/qt-6/supported-platforms.html) Windows 10 or above. There is also the recent [changes](https://www.qt.io/blog/qt-offering-changes-2020) where installing Qt now requires a Qt Account and offline installer is only available for commercial customers.

P.S. Qt is *huge*. I recently tried building it from source, ended up with ~115GB of files. Using the installer and selecting only a few modules still takes up more than 10GB.

## [Electron](https://www.electronjs.org/)

Used by many high-profile applications, infamous for being a resource hog. Being able to use web technologies is great, but it comes with a heavy price. I'd rather not go this route, though we could still consider using relatively lighter alternatives such as [Sciter](https://sciter.com/) or [Ultralight](https://ultralig.ht/).

## [WinUI 3](https://docs.microsoft.com/en-us/windows/apps/winui/winui3/)

This framework is what Microsoft offers for modernizing native Win32 desktop apps. You get a modern look (i.e. [Fluent Design System](https://docs.microsoft.com/en-us/windows/apps/design/)) while staying native. This is what you see in Windows 11's Settings window, for instance. Has many useful [controls](https://www.microsoft.com/en-us/p/winui-3-controls-gallery/9p3jfpwwdzrc), can instantly switch to dark mode. Biggest downside is that choosing this option would mean Taiga remains Windows-only (specifically, Windows 10 or above).

## Custom UI

We could always draw our own UI with something like [Dear ImGui](https://github.com/ocornut/imgui) or [NanoGUI](https://github.com/mitsuba-renderer/nanogui). At this point, Taiga would no longer look or behave like a desktop application. Probably not a good idea, mentioned here for completion's sake.

**Comment by vktr** [comment]
:wave: We have similar discussions in PicoTorrent as well - I can't say I have a good answer, just that I've been comparing mostly the same frameworks with mostly the same conclusions :smiley: 

One thing I'm currently experimenting with is building directly on top of [WebView2](https://docs.microsoft.com/en-us/microsoft-edge/webview2/). It means you can use HTML and CSS to design the user interface (but no external deps since WebView2 is bundled with Edge Chromium) and still use native code for things such as the BitTorrent things (in our case).

The app does take a bit more memory to run, and it runs in multiple processes due to how Chromium works, but it's a trade-off I'm mostly willing to make.

You can then use some glue code to build similar web views on other platforms with their respective bundled web view control.

**Comment by Jiyuu** [comment]
Just wanted to chime in with the mention that, assuming I got it right, https://platform.uno/how-it-works/ suggests WinUI3 might be cross-platform compatible.
Looking at https://platform.uno/blog/a-piece-of-windows-10-is-now-running-on-webassembly-natively-on-ios-and-android/ seems to suggest it can work with C++ codebases, to a degree.

Not sure if applicable but I saw the notification and figured it might be worth mentioning..

**Reply by somerendoguy** [reply]
Github rep: https://github.com/unoplatform/uno

**Comment by ghost** [comment]
I vote against Electron.
also don't forget [GTK](https://www.gtk.org/)

**Reply by ghost** [reply]
I have more up votes then OP :relaxed: , 
does this mean GTK is the winner? @erengy

**Reply by Bartek200219** [reply]
I think it's just about being against Electron

**Comment by QEDeD** [comment]
Will follow the progress with interest :+1:

**Comment by ifonefox** [comment]
wxWidgets recently improved their dark mode detection, by checking if it's on or off in the registry. Firefox and Chrome do it the same way. https://github.com/wxWidgets/wxWidgets/commit/622a2f9d6022d5ca2f1d76e93b2587ac12c2aa82

**Comment by kdbcinco** [comment]
Would you consider something like [React Native for Windows/Mac](https://microsoft.github.io/react-native-windows/)? I think it would also open up the possibility of having a mobile version of Taiga in the future. Another option would be [Flutter for Desktop](https://flutter.dev/multi-platform/desktop) but I'm a bit biased towards React Native since that's the one I'm using currently.

**Comment by alealclag** [comment]
Seems like wxWidgets is the best option, doesn't it?

**Comment by smowtenshi** [comment]
What about developing Taiga on something like Ubuntu, for example? There are Qt packages available with just one command, no need to think about any installers.

**Comment by soredake** [comment]
https://github.com/miurahr/aqtinstall

**Reply by erengy** [reply]
This seems to work better than the official installer. Thanks!

**Comment by ghost** [comment]
Electron would be awesome, i open the app only when i watch anime anyways so i dont see how high memory usage would be a problem 😄

**Reply by Bartek200219** [reply]
I have taiga always open so it can send notifications when new episode is available and high memory usage would be a problem.

**Comment by moharifhidayat7** [comment]


**Reply by smowtenshi** [reply]
That doesn't make sense if Taiga should go cross-platform route

**Comment by erengy** [comment]
After checking out most of our options, evaluating them in terms of features, appearance, developer experience, quality of shipped product and weighing their relevance for Taiga, I ended up with the following ordering:

    Qt > webview >= wxWidgets > WinUI > Custom

While there's no perfect match, given the tradeoffs and Taiga's priorities, Qt seems to be the best choice overall. I believe that its shortcomings are either avoidable or relatively less important. This is not final, but v2 will most likely be developed with Qt.

Thank you everyone for chiming in!

**Reply by erengy** [reply]
In case anyone's wondering, I did start developing v2 with Qt.

**Comment by ExileFox** [comment]
A little late to the discussion, but there are a few other options:
https://github.com/Immediate-Mode-UI/Nuklear
https://github.com/ocornut/imgui

Both are cross-platform and resource-efficient.

**Comment by suasgn** [comment]
https://gpui.rs/ from Zed might be an interesting option
